client/protocols/guidance.py

```diff
@@ -0,0 +1,178 @@
+        self.debug = debug
+        try:
+            with open(working_dir + OK_GUIDANCE,"r") as f:
+                self.guidance_json = json.load(f)
+            self.error = False
+        except:
+            self.error = True
+
+    def guidance_msg(self,unique_id, input_lines,access_token,hash_key,guidance_flag):
+        if self.error:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        shorten_unique_id = unique_id[unique_id.index('\n')+1:]
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        dict_info = wa_2_dict_info.get(repr(input_lines))
+
+        # If this wrong answer is not in the JSON file, display default message
+        if not dict_info:            
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+        if(self.debug != -1):
+            self.tg_id = self.debug
+        else:
+            self.get_tg(access_token,guidance_flag)
+
+        if self.tg_id == -1: 
+            # If self.tg_id == -1, some errors happen when trying to access server
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_string_key = self.guidance_json['dictTg2Func'].get(str(self.tg_id))
+
+        if not lambda_string_key:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
+
+        lst_mis_u = dict_info.get('lstMisU')
+
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        self.misU_count = self.update_misUcounts(hash_key, lst_mis_u, repr(input_lines),shorten_unique_id)
+
+        thresold = self.guidance_json['wrongAnsThresh']
+
+        """
+        Update count file using the helper. Each misunderstanding igets one count.
+
+        Get the updated acculamted counts as a dictionary: key: mis_u, value: counts.
+        
+        Called that dictionary self.misU_count
+        """
+
+        self.msg_id_set = set()
+
+        # If the count is higher than the thresold, we need to display the message
+        for mis_u in self.misU_count:
+            if self.misU_count[mis_u] >= thresold:
+                # Print each associated unique message for each miunderstanding
+                self.msg_id = lambda_info_misu(dict_info, mis_u)
+
+                if self.msg_id:
+                    self.msg_id_set.add(self.msg_id)
+
+        if len(self.msg_id_set) == 0 or None in self.msg_id_set:
+            # if student is in control group, just print the default message
```

Comment:
No, if `None` is in `msg_id_set` that does not necessarily mean they are in the control group. They are in the control group if that is what their number says.